File: rust/src/writer.rs
//! GolfWriter: append records, then seal into an immutable .golf file.

use crate::block::{compress_block, serialize_block};
use crate::header::{
    encode_metadata, write_header, Compression, Header, MetadataEntry, TimestampResolution,
};
use crate::index::{encode_index, BlockDescriptor};
use crate::{FOOTER_MAGIC, FORMAT_VERSION, FOOTER_SIZE};
use std::io::{self, BufWriter, Write};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum WriterError {
    #[error("record value size mismatch: expected {expected}, got {actual}")]
    ValueSizeMismatch { expected: u32, actual: usize },
    #[error("writer already sealed")]
    AlreadySealed,
    #[error("no records to seal")]
    EmptyFile,
    #[error("I/O error: {0}")]
    Io(#[from] io::Error),
    #[error("block compression error: {0}")]
    Block(#[from] crate::block::BlockError),
}

/// Configuration for building a golf file.
#[derive(Debug, Clone)]
pub struct WriterConfig {
    pub record_value_size: u32,
    pub ts_resolution: TimestampResolution,
    pub compression: Compression,
    pub block_capacity: u32,
    pub metadata: Vec<MetadataEntry>,
}

impl Default for WriterConfig {
    fn default() -> Self {
        WriterConfig {
            record_value_size: 64,
            ts_resolution: TimestampResolution::Nanoseconds,
            compression: Compression::Lz4,
            block_capacity: 8192,
            metadata: Vec::new(),
        }
    }
}

/// A writer that accumulates records and seals them into a .golf file.
pub struct GolfWriter {
    config: WriterConfig,
    records: Vec<(u64, Vec<u8>)>,
    sealed: bool,
}

impl GolfWriter {
    /// Create a new writer with the given configuration.
    pub fn new(config: WriterConfig) -> Self {
        GolfWriter {
            config,
            records: Vec::new(),
            sealed: false,
        }
    }

    /// Append a record. The value must be exactly `record_value_size` bytes.
    pub fn append(&mut self, timestamp: u64, value: Vec<u8>) -> Result<(), WriterError> {
        if self.sealed {
            return Err(WriterError::AlreadySealed);
        }
        if value.len() != self.config.record_value_size as usize {
            return Err(WriterError::ValueSizeMismatch {
                expected: self.config.record_value_size,
                actual: value.len(),
            });
        }
        self.records.push((timestamp, value));
        Ok(())
    }

    /// Return the number of records appended so far.
    pub fn len(&self) -> usize {
        self.records.len()
    }

    /// Return whether the writer has no records.
    pub fn is_empty(&self) -> bool {
        self.records.is_empty()
    }

    /// Seal the writer and write the .golf file to the given writer.
    /// Consumes self.
    pub fn seal<W: Write>(mut self, dest: W) -> Result<(), WriterError> {
        if self.sealed {
            return Err(WriterError::AlreadySealed);
        }
        if self.records.is_empty() {
            return Err(WriterError::EmptyFile);
        }
        self.sealed = true;

        // Sort records by timestamp
        self.records.sort_by_key(|(ts, _)| *ts);

        let mut w = BufWriter::new(dest);
        let record_value_size = self.config.record_value_size;
        let block_capacity = self.config.block_capacity as usize;

        // Encode metadata
        let meta_bytes = encode_metadata(&self.config.metadata);

        // Compute min/max timestamps
        let min_ts = self.records.first().unwrap().0;
        let max_ts = self.records.last().unwrap().0;

        // Build header
        let header = Header {
            version: FORMAT_VERSION,
            flags: 0,
            record_value_size,
            ts_resolution: self.config.ts_resolution,
            compression: self.config.compression,
            block_capacity: self.config.block_capacity,
            min_timestamp: min_ts,
            max_timestamp: max_ts,
            record_count: self.records.len() as u64,
            metadata_length: meta_bytes.len() as u32,
        };

        // Write header
        write_header(&mut w, &header, &self.config.metadata)?;

        // Track current file offset
        let mut offset = 64u64 + meta_bytes.len() as u64;

        // Build and write blocks
        let mut descriptors = Vec::new();
        let chunks: Vec<_> = self.records.chunks(block_capacity).collect();
        for chunk in &chunks {
            // Build the raw record pairs as (u64, &[u8])
            let record_refs: Vec<(u64, &[u8])> =
                chunk.iter().map(|(ts, v)| (*ts, v.as_slice())).collect();

            // Serialize block
            let raw = serialize_block(&record_refs, record_value_size);
            let block_crc = crc32c::crc32c(&raw);
            let uncompressed_size = raw.len() as u32;

            // Compress
            let compressed = compress_block(&raw, self.config.compression)?;
            let compressed_size = compressed.len() as u32;

            // Record descriptor
            let min_block_ts = chunk.first().unwrap().0;
            let max_block_ts = chunk.last().unwrap().0;
            descriptors.push(BlockDescriptor {
                min_ts: min_block_ts,
                max_ts: max_block_ts,
                block_offset: offset,
                compressed_size,
                uncompressed_size,
                record_count: chunk.len() as u32,
                block_crc,
            });

            // Write block
            w.write_all(&compressed)?;
            offset += compressed_size as u64;
        }

        // Write block index
        let index_offset = offset;
        let index_bytes = encode_index(&descriptors);
        let index_crc = crc32c::crc32c(&index_bytes);
        w.write_all(&index_bytes)?;

        // Write footer
        let block_count = descriptors.len() as u64;
        let mut footer = [0u8; FOOTER_SIZE];
        footer[0..8].copy_from_slice(&index_offset.to_le_bytes());
        footer[8..16].copy_from_slice(&block_count.to_le_bytes());
        footer[16..20].copy_from_slice(&index_crc.to_le_bytes());
        footer[20..24].copy_from_slice(&FOOTER_MAGIC);
        let footer_crc = crc32c::crc32c(&footer[0..24]);
        footer[24..28].copy_from_slice(&footer_crc.to_le_bytes());
        // 28..32 padding (zeros)
        w.write_all(&footer)?;
        w.flush()?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_writer_value_size_mismatch() {
        let mut w = GolfWriter::new(WriterConfig {
            record_value_size: 8,
            ..Default::default()
        });
        let result = w.append(1, vec![0u8; 16]);
        assert!(result.is_err());
    }

    #[test]
    fn test_writer_seal_empty() {
        let w = GolfWriter::new(WriterConfig::default());
        let mut buf = Vec::new();
        let result = w.seal(&mut buf);
        assert!(result.is_err());
    }

    #[test]
    fn test_writer_basic_seal() {
        let mut w = GolfWriter::new(WriterConfig {
            record_value_size: 8,
            compression: Compression::None,
            block_capacity: 4,
            ..Default::default()
        });
        for i in 0..10u64 {
            w.append(i * 100, vec![i as u8; 8]).unwrap();
        }
        let mut buf = Vec::new();
        w.seal(&mut buf).unwrap();
        // Should have valid header magic
        assert_eq!(&buf[0..4], b"GOLF");
        // Should have valid footer magic
        let footer_start = buf.len() - FOOTER_SIZE;
        assert_eq!(&buf[footer_start + 20..footer_start + 24], b"FLOG");
    }
}

File: swift/Sources/Golf/Writer.swift
// Writer.swift — append-then-seal construction of .golf files.

import Foundation

/// Configuration for ``GolfWriter``.
public struct WriterConfig: Sendable {
    /// Required size in bytes of every appended record's value.
    public var recordValueSize: Int
    /// Unit of the timestamps you will append. Default `.nanoseconds`.
    public var tsResolution: TimestampResolution
    /// Block codec. Default `.lz4`.
    public var compression: Compression
    /// Maximum records per data block (hence per index entry). Default `8192`.
    /// Larger blocks compress better; smaller blocks make narrow range queries cheaper.
    public var blockCapacity: Int
    /// User-defined key/value pairs embedded in the header.
    public var metadata: [MetadataEntry]

    public init(
        recordValueSize: Int,
        tsResolution: TimestampResolution = .nanoseconds,
        compression: Compression = .lz4,
        blockCapacity: Int = 8192,
        metadata: [MetadataEntry] = []
    ) {
        self.recordValueSize = recordValueSize
        self.tsResolution = tsResolution
        self.compression = compression
        self.blockCapacity = blockCapacity
        self.metadata = metadata
    }
}

/// Accumulates `(timestamp, value)` records and seals them into an immutable
/// golf file.
///
/// Usage: configure once, `append` any number of records in any order, then
/// call ``seal()`` exactly once. Sealing sorts by timestamp, partitions into
/// blocks of up to ``WriterConfig/blockCapacity``, compresses each block
/// independently, writes the block index and footer, and returns the whole
/// file as a byte buffer.
public final class GolfWriter {
    private let config: WriterConfig
    private var records: [Record] = []
    private var sealed = false

    /// Number of records buffered so far.
    public var count: Int { records.count }

    /// - Parameter config: see ``WriterConfig``.
    public init(config: WriterConfig) {
        self.config = config
    }

    /// Buffers one record. The value is copied and must be exactly
    /// `recordValueSize` bytes long.
    public func append(timestamp: UInt64, value: [UInt8]) throws {
        if sealed { throw GolfError.writerAlreadySealed }
        guard value.count == config.recordValueSize else {
            throw GolfError.valueSizeMismatch(expected: config.recordValueSize, actual: value.count)
        }
        records.append(Record(timestamp: timestamp, value: value))
    }

    /// Convenience that builds the value from a little-endian integer.
    ///
    /// The integer is serialized with shift-based byte extraction (endian-
    /// independent on any host), then zero-padded in the high-order bytes to
    /// `recordValueSize` (or truncated to its low 8 bytes when smaller).
    public func append(timestamp: UInt64, intValue: UInt64) throws {
        let size = max(config.recordValueSize, 0)
        let raw = intValue.bytes()
        if size > 8 {
            var padded = [UInt8](repeating: 0, count: size)
            padded.replaceSubrange(0..<8, with: raw)
            try append(timestamp: timestamp, value: padded)
        } else {
            try append(timestamp: timestamp, value: Array(raw[0..<min(8, size)]))
        }
    }

    /// Seals the accumulated records and returns the complete `.golf` file image.
    public func seal() throws -> [UInt8] {
        if sealed { throw GolfError.writerAlreadySealed }
        guard !records.isEmpty else { throw GolfError.noRecords }
        sealed = true

        // Spec: sort ascending; ties keep implementation-defined order.
        records.sort { $0.timestamp < $1.timestamp }

        let capacity = max(1, config.blockCapacity)
        let metaBytes = try encodeMetadata(config.metadata)

        var parts: [[UInt8]] = []
        parts.append(
            GolfHeader(
                recordValueSize: config.recordValueSize,
                tsResolution: config.tsResolution,
                compression: config.compression,
                blockCapacity: capacity,
                minTimestamp: records.first!.timestamp,
                maxTimestamp: records.last!.timestamp,
                recordCount: UInt64(records.count),
                metadataLength: metaBytes.count
            ).encode()
        )
        if !metaBytes.isEmpty { parts.append(metaBytes) }

        var offset = UInt64(Golf.fixedHeaderSize + metaBytes.count)
        var descriptors: [BlockDescriptor] = []

        var index = 0
        while index < records.count {
            let chunk = Array(records[index..<Swift.min(index + capacity, records.count)])
            let raw = serializeBlock(chunk, recordValueSize: config.recordValueSize)

            let compressed: [UInt8]
            switch config.compression {
            case .none:
                compressed = raw
            case .lz4:
                compressed = lePrefixed(uncompressedCount: raw.count,
                                        payload: Lz4Block.compress(raw))
            case .zstd:
                throw GolfError.compressionError("Zstd writing is not supported by this implementation")
            }

            descriptors.append(
                BlockDescriptor(
                    minTs: chunk.first!.timestamp,
                    maxTs: chunk.last!.timestamp,
                    blockOffset: offset,
                    compressedSize: compressed.count,
                    uncompressedSize: raw.count,
                    recordCount: chunk.count,
                    blockCrc: CRC32C.checksum(raw)
                )
            )
            parts.append(compressed)
            offset += UInt64(compressed.count)
            index += capacity
        }

        let indexOffset = offset
        let indexBytes = descriptors.flatMap { $0.encode() }
        parts.append(indexBytes)
        parts.append(
            Footer(
                indexOffset: indexOffset,
                blockCount: UInt64(descriptors.count),
                indexCrc: CRC32C.checksum(indexBytes)
            ).encode()
        )

        return parts.flatMap { $0 }
    }

    /// Seals directly to a file at `url`, replacing any existing file.
    public func seal(to url: URL) throws {
        let data = Data(try seal())
        try data.write(to: url, options: .atomic)
    }

    /// `[u32 LE uncompressed size][payload]` — the on-disk LZ4 convention
    /// shared with the reference implementations (`lz4_flex` prepend-size).
    private func lePrefixed(uncompressedCount: Int, payload: [UInt8]) -> [UInt8] {
        var out = [UInt8](repeating: 0, count: 4 + payload.count)
        LE.writeU32(UInt32(uncompressedCount), into: &out, 0)
        out.replaceSubrange(4..<out.count, with: payload)
        return out
    }
}

private extension UInt64 {
    func bytes() -> [UInt8] {
        (0..<8).map { UInt8((self >> (8 * $0)) & 0xFF) }
    }
}
